app/uk/gov/hmrc/transitmovements/services/ObjectStoreService.scala

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.transitmovements.services
+
+import akka.NotUsed
+import akka.stream.scaladsl.Source
+import akka.util.ByteString
+import cats.data.EitherT
+import com.google.inject.ImplementedBy
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.objectstore.client.play.PlayObjectStoreClient
+import uk.gov.hmrc.objectstore.client.play.Implicits._
+import uk.gov.hmrc.transitmovements.services.errors.ObjectStoreError
+import uk.gov.hmrc.objectstore.client.Path
+import uk.gov.hmrc.transitmovements.models.ObjectStoreURI
+
+import javax.inject._
+import scala.concurrent._
+import scala.util.control.NonFatal
+
+@ImplementedBy(classOf[ObjectStoreServiceImpl])
+trait ObjectStoreService {
+
+  def getObjectStoreFile(
+    objectStoreURI: ObjectStoreURI
+  )(implicit ec: ExecutionContext, hc: HeaderCarrier): EitherT[Future, ObjectStoreError, Source[ByteString, _]]
+
+}
+
+@Singleton
+class ObjectStoreServiceImpl @Inject() (client: PlayObjectStoreClient) extends ObjectStoreService {
+
+  override def getObjectStoreFile(
+    objectStoreURI: ObjectStoreURI
+  )(implicit ec: ExecutionContext, hc: HeaderCarrier): EitherT[Future, ObjectStoreError, Source[ByteString, _]] =
+    EitherT(
+      client
+        .getObject[Source[ByteString, NotUsed]](
+          Path.File(objectStoreURI.path.map(_.mkString).getOrElse(objectStoreURI.value)),
```

Comment:
If I send down `transit-movements-router/movements/abcdef0123456789/message/abcdef-20230701-120000.xml`, that's a mistake, right? That lets this through. Returning a `None` is an error condition so we should be treating it as such, right?

Reply:
ok. so do you want me to use like objectStoreURI.path.map(_.mkString) only? so sending transit-movements-router/movements/abcdef0123456789/message/abcdef-20230701-120000.xml will return None here.

Reply:
Nope. That's correct.

Reply:
This line still doesn't reflect what I asked, and it doesn't really reflect what you said. This line is calling `object-store` when we **know** the result will be wrong, which we should not be doing. If we can't extract the path because it's of the wrong form, we should fail fast and NOT call object-store just to get an error code.

Whatever is in the option returned back from `objectStoreURI.path` should be the ONLY value that we consider here. I really don't know why we have the `getOrElse` here just using the URI, because by out specification, that's always going to be wrong, so why then reach out to object-store to determine that, just to then return the error you could short circuit?

Perhaps consider breaking this function up a little and putting it behind a for comprehension if you're worried about trying to make this totally functional:

```scala
override def getObjectStoreFile(
    objectStoreURI: ObjectStoreURI
  )(implicit ec: ExecutionContext, hc: HeaderCarrier): EitherT[Future, ObjectStoreError, Source[ByteString, _]] =
    for {
        path   <- getPath(objectStoreURI)
        result <- getObjectFromObjectStore(path)
    } yield result

private def getPath(objectStoreUri: ObjectStoreURI)(implicit ec: ExecutionContext): EitherT[Future, ObjectStoreError, Source[ByteString, _]] =
    EitherT(Future.successful(objectStoreUri.path.toRight(ObjectStoreError.InvalidURI)))

private def getObjectFromObjectStore(path: String)(implicit ec: ExecutionContext, hc: HeaderCarrier): EitherT[Future, ObjectStoreError, Source[ByteString, _]] = ???
```

Also, to note:

* No need to call `mkString` on a string, which you do in `objectStoreURI.path.map(_.mkString)`
* You repeated `objectStoreURI.path.map(_.mkString).getOrElse(objectStoreURI.value)` below on line 58 -- be careful not to repeat yourself in these cases.

Reply:
ohk I got your point. In that case I think better to handle in it in ObjectStoreURIHeaderExtractor  instead of doing it in Service so that we can catch it while validating ObjectStoreURIHeader. Please suggest if below one is looks good
trait ObjectStoreURIHeaderExtractor {

  def extractObjectStoreURI(headers: Headers): EitherT[Future, HeaderExtractError, ObjectStoreURI] =
    EitherT {
      headers.get(Constants.ObjectStoreURI) match {
        case None => Future.successful(Left(NoHeaderFound("Missing X-Object-Store-Uri header value")))
        case Some(headerValue) =>
          ObjectStoreURI(headerValue).path match {
            case None                 => Future.successful(Left(InvalidObjectStoreURI(s"Invalid X-Object-Store-Uri header value: $headerValue")))
            case Some(objectStoreURI) => Future.successful(Right(ObjectStoreURI(objectStoreURI)))
          }
      }
    }

Reply:
If you like, but I'd rather not have nested pattern matching if we can help it.